File: Solutions/1026_Maximum_Difference_Between_Node_and_Ancestor/1026_Maximum_Difference_Between_Node_and_Ancestor.go
//Second fastest
/*
To make the difference between x and y bigger, we either:

1. make x greater, when x >= y
2. make x smaller, when x < y
So we can pass the smallest and largest nodes along with the process of DFS
*/

func maxAncestorDiff(root *TreeNode) int {
	sol := Solution{}
	return sol.Run(root)
}

type Solution struct {
	maxDiff int
}

func (s *Solution) Run(root *TreeNode) int {
	s.helper(root, root.Val, root.Val)
	return s.maxDiff
}

func (s *Solution) helper(root *TreeNode, minAnc, maxAnc int) {
	if root == nil {
		return
	}
	// get diff
	d1, d2 := math.Abs(float64(minAnc-root.Val)), math.Abs(float64(maxAnc-root.Val))
	// max diff
	d := math.Max(d1, d2)
	s.maxDiff = int(math.Max(d, float64(s.maxDiff)))

	// try to get the minimum and maximum ancestors, which could give us the best chances of getting maximum difference
	minAnc, maxAnc = int(math.Min(float64(minAnc), float64(root.Val))), int(math.Max(float64(maxAnc), float64(root.Val)))
	s.helper(root.Left, minAnc, maxAnc)
	s.helper(root.Right, minAnc, maxAnc)
}

//Another
func maxAncestorDiff(root *TreeNode) int {
	if root == nil {
		return 0
	}
	var tracks []int
	// reMax := recursion(root, &tracks, 0)
	// itMax := iteration(root, &tracks, 0)
	itMax := iterationMaxMin(root, &tracks, 0)

	return itMax
}

func abs(num int) int {
	if num < 0 {
		return -num
	}
	return num
}

func max(num1, num2 int) int {
	if num1 > num2 {
		return num1
	}
	return num2
}

func iterationMaxMin(root *TreeNode, tracks *[]int, itMax int) int {
	type wrapper struct {
		node    *TreeNode
		visited int
	}
	var stack []*wrapper
	stack = append(stack, &wrapper{root, 0})
	var top *wrapper
	anceMax, anceMin := root.Val, root.Val
	for len(stack) > 0 {
		top = stack[len(stack)-1]
		if top.node == nil {
			stack = stack[:len(stack)-1]
			continue
		}

		if top.visited == 1 {
			*tracks = (*tracks)[:len(*tracks)-1]
			stack = stack[:len(stack)-1]
			if top.node.Val == anceMin {
				temp := 1000000
				for _, val := range *tracks {
					if val < temp {
						temp = val
					}
				}
				anceMin = temp
			}
			if top.node.Val == anceMax {
				temp := 0
				for _, val := range *tracks {
					if val > temp {
						temp = val
					}
				}
				anceMax = temp
			}
			continue
		}

		if top.node.Val < anceMin {
			anceMin = top.node.Val
		}
		if top.node.Val > anceMax {
			anceMax = top.node.Val
		}
		if abs(anceMax-anceMin) > itMax {
			itMax = abs(anceMax - anceMin)
		}

		top.visited = 1
		*tracks = append(*tracks, top.node.Val)
		stack = append(stack, &wrapper{top.node.Right, 0})
		stack = append(stack, &wrapper{top.node.Left, 0})
	}
	return itMax
}


//Fastest
func maxAncestorDiff(root *TreeNode) int {
    largest := 0
    helper(root, &largest)
    return largest
}

func helper(node *TreeNode, largest *int) (int, int) {
    if node.Left == nil && node.Right == nil {
        return node.Val, node.Val
    }
    
    leftMax, leftMin := node.Val, node.Val
    if node.Left != nil {
        leftMax, leftMin = helper(node.Left, largest)
    }
    
    rightMax, rightMin := node.Val, node.Val
    if node.Right != nil {
        rightMax, rightMin = helper(node.Right, largest)
    }
    
    
    localMax := 0
    if abs(leftMax - node.Val) > abs(leftMin - node.Val) {
        localMax = abs(leftMax - node.Val)
    } else {
        localMax = abs(leftMin - node.Val)
    }
    
    if localMax > *largest {
        *largest = localMax
    }
    
    if abs(rightMax - node.Val) > abs(rightMin - node.Val) {
        localMax = abs(rightMax - node.Val)
    } else {
        localMax = abs(rightMin - node.Val)
    }
    
    if localMax > *largest {
        *largest = localMax
    }
    
    return max(rightMax, leftMax, node.Val), min(rightMin, leftMin, node.Val)
    
}

func abs(x int) int {
    if x > 0 {
        return x
    }
    return -x
}

func min(a, b, c int) int {
    smallest := a
    if b < smallest {
        smallest = b
    }
    
    if c < smallest {
        smallest = c
    }
    
    return smallest
}

func max(a, b, c int) int {
    biggest := a
    if b > biggest {
        biggest = b
    }
    
    if c > biggest {
        biggest = c
    }
    
    return biggest
    
}

File: Solutions/1026_Maximum_Difference_Between_Node_and_Ancestor/1026_Maximum_Difference_Between_Node_and_Ancestor.js
var maxAncestorDiff = function (root) {
    let result = 0;
    function helper(root, max, min) {
        if (!root) {
            if (result < max - min) {
                result = max - min;
            }
            return;
        }
        if (root.val > max) {
            max = root.val
        }
        if (root.val < min) {
            min = root.val
        }
        helper(root.left, max, min);
        helper(root.right, max, min);
    }
    helper(root, root.val, root.val);
    return result;
};

//Third fastest
var maxAncestorDiff = function(root) {
    if(!root) return 0;
    let result = 0;
    
    const dfs = (root, min, max) => {
        if(!root){
            result = Math.max(result, max - min);
        } else {
            max = Math.max(root.val, max);
            min = Math.min(root.val, min);
            dfs(root.left, min, max);
            dfs(root.right, min, max);
        }
    };
    
    dfs(root, Infinity, -Infinity);
    
    return result;
};

//Second fastest
var maxAncestorDiff = function(root) {
    
    var rst = -1;
    if(root) traverse(root);
    return rst;
    
    function traverse(root){
        let [min,max,lmin,lmax]=[root.val,root.val,root.val,root.val];
        if(root.left){
            [lmin,lmax]=traverse(root.left);
            min = Math.min(min,lmin);
            max = Math.max(max,lmax);
        } 
        if(root.right){
            [lmin,lmax]=traverse(root.right);
            min = Math.min(min,lmin);
            max = Math.max(max,lmax);
        }
        rst = Math.max(rst,Math.abs(root.val-min),Math.abs(root.val-max));
        return [min,max];
    }
};

//Fastest
var maxAncestorDiff = function(root) {
    let diff = -Infinity;
    
    let iterate = (root, max, min) => {
        if(!root) {
            if(diff < max - min) {
                diff = max - min;
            }
            return;
        }

        if(root.val > max) {
            max = root.val
        }
        
        if(root.val < min) {
            min = root.val
        }
        
        iterate(root.left, max, min)
        iterate(root.right, max, min)
    }
    
    iterate(root, root.val, root.val);
    
    return diff;
};